Makefile

```diff
@@ -178,31 +178,34 @@ cfn/lint: | guard/program/cfn-lint
 ## Runs eclint against the project
 eclint/lint: | guard/program/eclint guard/program/git
 eclint/lint: HAS_UNTRACKED_CHANGES ?= $(shell cd $(PROJECT_ROOT) && git status -s)
-eclint/lint: ECLINT_FILES ?= git ls-files -z
+eclint/lint: ECLINT_FILES ?= $(shell git -C $(PROJECT_ROOT) ls-files -z | grep -zv ".bats" | xargs -0)
 eclint/lint:
 	@ echo "[$@]: Running eclint..."
 	cd $(PROJECT_ROOT) && \
 	[ -z "$(HAS_UNTRACKED_CHANGES)" ] || (echo "untracked changes detected!" && exit 1)
-	cd $(PROJECT_ROOT) && \
-	$(ECLINT_FILES) | grep -zv ".bats" | xargs -0 -I {} eclint check {}
+	eclint check $(ECLINT_FILES)
 	@ echo "[$@]: Project PASSED eclint test!"
 
-python/%: PYTHON_FILES ?= git ls-files --cached --others --exclude-standard '*.py'
+python/%: PYTHON_FILES ?= $(shell git -C $(PROJECT_ROOT) ls-files --cached --others --exclude-standard '*.py' | xargs printf "$(PROJECT_ROOT)/%s ")
 ## Checks format and lints Python files.  Runs pylint on each individual
 ## file and uses a custom format for the lint messages.
 python/lint: | guard/program/pylint guard/program/black guard/program/git
 python/lint:
 	@ echo "[$@]: Linting Python files..."
-	$(PYTHON_FILES) | xargs black --check
-	$(PYTHON_FILES) | $(XARGS) -n1 pylint -rn -sn \
-		--msg-template="{path}:{line} [{symbol}] {msg}" {}
-	@ echo "[$@]: Python files PASSED lint test!"
+	black --check $(PYTHON_FILES)
+	for python_file in $(PYTHON_FILES); do \
+		if [[ $$python_file = *.py ]]; then \
```

Comment:
can we remove this check? we're already doing this test in the default PYTHON_FILES command, and if the user provides a bad command then we should error. they can then update their command to filter the list.

Reply:
This check does give the user some flexibility, but it also accommodates the case where there are no Python files in the directory.  When there is an empty list of Python files, the 'printf' still appends $PROJECT_ROOT to an empty argument, i.e., PYTHON_FILES=$PROJECT_ROOT.  It's ugly and I tried various ways to get around it.  This logic was clear and allowed for users to inadvertently provide use with non-Python files and for us to silently ignore those files.

Reply:
Silently ignoring failures is just about the root of all evil in shell programs.

How about this, using `xargs --no-run-if-empty`?

```
python/%: PYTHON_FILES ?= $(shell git -C $(PROJECT_ROOT) ls-files --cached --others --exclude-standard '*.py' | xargs --no-run-if-empty printf "$(PROJECT_ROOT)/%s ")
```

testing locally...
```
$ git ls-files --cached --others --exclude-standard '*.py' | xargs printf "foo/%s "
foo/
$ git ls-files --cached --others --exclude-standard '*.py' | xargs --no-run-if-empty printf "foo/%s "
$
```

Reply:
That xargs argument is not supported on Macs.

Reply:
Also, it's not really a failure.  I could echo a statement indicating that a non-python file was being skipped for a python lint function, but then I'd had to test for the case of $PROJECT_ROOT.

Reply:
On Macs, pretty much everyone seems to install the gnutils version of xargs (using `brew install findutils`), to get an actual modern version of the utility. Then it works fine. We should have no problem making that a requirement for this project.

Reply:
The .py extension is a convention, not a requirement. Any file can be a valid python file. This is actually one of the reasons the caller may want to override our default PYTHON_FILES.

```
$ cat test.foo
print("Hello World!")

$ python test.foo
Hello World!
```

Reply:
I wish I had know that about the macs.  That would have saved me a lot of time.  I'll make the change.